tests/results_benchmark.py

```diff
@@ -137,62 +176,170 @@ def __init__(self, root: Path = Path("data/raw/pipal")):
+
+    return distorted_features, reference_features
+
+
+def normalize_tensor(tensor: torch.Tensor) -> torch.Tensor:
+    """ Map tensor values to [0, 1] """
+    return (tensor - tensor.min()) / (tensor.max() - tensor.min())
+
+
+def compute_distribution_based(metric_functor: Callable, distorted_images: torch.Tensor,
+                               reference_images: torch.Tensor, device: str, feature_extractor_name: str) -> np.ndarray:
+    feature_extractor = get_feature_extractor(feature_extractor_name=feature_extractor_name, device=device)
+
+    if feature_extractor_name == 'inception':
+        distorted_images = normalize_tensor(distorted_images)
+        reference_images = normalize_tensor(reference_images)
+
+    # Create patches
+    distorted_patches = crop_patches(distorted_images, size=96, stride=32)
+    reference_patches = crop_patches(reference_images, size=96, stride=32)
+
+    # Extract features from distorted images
+    distorted_patches = distorted_patches.view(-1, *distorted_patches.shape[-3:])
+    reference_patches = reference_patches.view(-1, *reference_patches.shape[-3:])
+
+    distorted_features, reference_features = extract_features(distorted_patches, feature_extractor,
+                                                              feature_extractor_name, reference_patches)
+
+    return metric_functor(distorted_features, reference_features).cpu()
+
+
+def crop_patches(images: torch.Tensor, size=64, stride=32):
+    """Crop input images into smaller patches
+    Args:
+        images: Tensor of images with shape (batch x 3 x H x W)
+        size: size of a square patch
+        stride: Step between patches
+    """
+    patches = images.data.unfold(1, 3, 3).unfold(2, size, stride).unfold(3, size, stride)
+    patches = patches.reshape(-1, 3, size, size)
+    return patches
+
+
+def main(dataset_name: str, path: Path, metrics: List[str], batch_size: int, device: str, feature_extractor: str) \
+        -> None:
     # Init dataset and dataloader
     dataset = DATASETS[dataset_name](root=path)
-    loader = torch.utils.data.DataLoader(dataset, batch_size=batch_size, num_workers=0)
+    loader = DataLoader(dataset, batch_size=batch_size, num_workers=4)
+
+    # If category of metrics is selected instead of a list of metrics, take all metrics from this category
+    if metrics[0] in METRIC_CATEGORIES:
+        metrics = METRIC_CATEGORIES[metrics[0]]
 
-    for name in metrics:
-        gt_scores, metric_scores = eval_metric(loader, METRICS[name], device=device)
-        print(f"{name}: SRCC {abs(spearmanr(gt_scores, metric_scores)[0]):0.3f}",
+    if metrics[0] == 'all':
+        metrics = list(chain(*METRIC_CATEGORIES.values()))
+
+    for metric_name in metrics:
+        metric: Metric = METRICS[metric_name]
+        gt_scores, metric_scores = eval_metric(loader, metric, device=device, feature_extractor=feature_extractor)
+        print(f"{metric_name}: SRCC {abs(spearmanr(gt_scores, metric_scores)[0]):0.3f}",
```

Comment:
Good to have feature but not necessary one. Would it be possible to log the results into a file to minimise risks of losing outputs in stdout?

Reply:
I would keep the code minimal. In practice, logging into a file is as simple as running the script with the following line:  
`$ python results_benchmark.py >> results.txt`